src/Orleans/Serialization/SerializationManager.cs

```diff
@@ -1366,73 +1366,89 @@ public static T DeserializeInner<T>(BinaryTokenStreamReader stream)
         /// <returns>Object of the required Type, rehydrated from the input stream.</returns>
         public static object DeserializeInner(Type expected, BinaryTokenStreamReader stream)
         {
-            var start = stream.CurrentPosition;
+            var previousOffset = DeserializationContext.Current.CurrentObjectOffset;
+            DeserializationContext.Current.CurrentObjectOffset = stream.CurrentPosition;
 
-            // NOTE: we don't check that the actual dynamic result implements the expected type. We'll allow a cast exception higher up to catch this.
+            try
```

Comment:
Most of this code is unchanged, but I wrapped it all in try/finally to ensure that `DeserializationContext.Current.CurrentObjectOffset` is reset.